### PR TITLE
[BugFix]fix prepare execute bug when use jdbc 

### PR DESCRIPTION
**Pull request by liubotao**
## Why I'm doing:

When I was testing the Prepare function, the code threw an NPE error

```
----starRocksSelect----
customer_key:1, customer_row_value_0:a, customer_row_value_1:a, customer_row_value_2:a
java.sql.SQLSyntaxErrorException: NullPointerException, msg: null
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.ServerPreparedStatement.serverExecute(ServerPreparedStatement.java:637)
	at com.mysql.cj.jdbc.ServerPreparedStatement.executeInternal(ServerPreparedStatement.java:418)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeQuery(ClientPreparedStatement.java:1003)
	at ToolApplication.starRocksSelect(ToolApplication.java:84)
	at ToolApplication.main(ToolApplication.java:162)
```

Prepare Test Code 

```
public static void starRocksSelect() {
      String url = "jdbc:mysql://xxxxx:9030/demo?useServerPrepStmts=true";
      String user = "xxxx";
      String password = "xxxxx";

      try {
          Class.forName("com.mysql.cj.jdbc.Driver");
          Connection conn = DriverManager.getConnection(url, user, password);
          String query = "SELECT customer_key, customer_row_value_0, customer_row_value_1, customer_row_value_2 FROM customer_row WHERE customer_key = ?";
          PreparedStatement pstmt = conn.prepareStatement(query);
          pstmt.setInt(1, 1); 
          ResultSet rs = pstmt.executeQuery(); 
          
          while (rs.next()) {
              BigDecimal c0 = rs.getBigDecimal("customer_key");
              String c1 = rs.getString("customer_row_value_0");
              String c2 = rs.getString("customer_row_value_1");
              String c3 = rs.getString("customer_row_value_2");
              System.out.println("customer_key:" + c0 + ", customer_row_value_0:" + c1 + ", customer_row_value_1:" + c2 + ", customer_row_value_2:" + c3);
          }
          rs.close(); 
          
          pstmt.setInt(1, 2); 
          rs = pstmt.executeQuery(); 
          
          while (rs.next()) {
              BigDecimal c0 = rs.getBigDecimal("customer_key");
              String c1 = rs.getString("customer_row_value_0");
              String c2 = rs.getString("customer_row_value_1");
              String c3 = rs.getString("customer_row_value_2");
              System.out.println("customer_key:" + c0 + ", customer_row_value_0:" + c1 + ", customer_row_value_1:" + c2 + ", customer_row_value_2:" + c3);
          }
          rs.close();
          
          pstmt.close();
          conn.close();

      } catch (ClassNotFoundException | SQLException e) {
          e.printStackTrace();
      }
  }

  public static void main(String[] args) {
      starRocksSelect();
  }
```

customer_row table structure

```
CREATE TABLE `customer_row` (
  `customer_key` bigint(20) NOT NULL COMMENT "",
  `customer_row_value_0` varchar(65533) NULL COMMENT "",
  `customer_row_value_1` varchar(65533) NULL COMMENT "",
  `customer_row_value_2` varchar(65533) NULL COMMENT ""
) ENGINE=OLAP
PRIMARY KEY(`customer_key`)
DISTRIBUTED BY HASH(`customer_key`)
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"enable_persistent_index" = "true",
"replicated_storage" = "true",
"storage_type" = "COLUMN_WITH_ROW",
"compression" = "LZ4"
);
```

## What I'm doing:

Adjust the placeholder Params MysqlTypeCodes in the handleExecute method to the PrepareStmt in the Session preparedStmtCtx

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java [hunk 1]
```diff
@@ -532,7 +532,8 @@ private void handleExecute() {
             // new_params_bind_flag
             if (packetBuf.hasRemaining() && (int) packetBuf.get() != 0) {
                 // parse params types
-                IntStream.range(0, numParams).forEach(i -> mysqlTypeCodes[i] = (int) packetBuf.getChar());
+                IntStream.range(0, numParams)
+                        .forEach(i -> prepareCtx.getStmt().addMysqlTypeCodes((int) packetBuf.getChar()));
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java:536]: Why not use `Integer[] mysqlTypeCodes = new Integer[numParams];`?  `mysqlTypeCodes ` is no used and can be removed?

Add some comments plz.

Comment (liubotao) [reply fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java:536]: Integer[] mysqlTypeCodes = new Integer[numParams] Should be removed

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java:535]: Why add elements into the same `mysqlTypeCodes` in `PrepareStmt`?  Is there any problem lead the list size larger than placeholder number?

Comment (liubotao) [reply fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java:535]: When executing the first execute, the packetBuf[18] is {[23, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 3, 0, 1, 0, 0, 0]}.

When executing the second execute, the packetBuf[16] is {[23, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 2, 0, 0, 0]}.

During the execution of the second execute, based on the content of the packetBuf package, the condition packetBuf.get() != 0 cannot be used to set mysqlTypeCodes. It is necessary to reuse the existing mysqlTypeCodes placeholders in the session prepareContextStmt

Comment (liubotao) [reply fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java:535]: > Why add elements into the same `mysqlTypeCodes` in `PrepareStmt`? Is there any problem lead the list size larger than placeholder number?

Yes, I'll take care of it.

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/PrepareStmt.java [hunk 2]
```diff
@@ -74,6 +79,14 @@ public List<String> getParameterLabels() {
         return labels;
     }
 
+    public void addMysqlTypeCodes(Integer mysqlTypeCode) {
+        mysqlTypeCodes.add(mysqlTypeCode);
+    }
+
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/PrepareStmt.java:85]: please also remove this code.

**Review by packy92 — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/PrepareStmt.java (hunk 2 above)
Comment (liubotao) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/PrepareStmt.java:85]: Ok～

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/PrepareStmt.java [hunk 3]
```diff
@@ -31,11 +31,18 @@ public class PrepareStmt extends StatementBase {
...
+    protected List<Integer> mysqlTypeCodes = new ArrayList<>();
+
     public PrepareStmt(String name, StatementBase stmt, List<Parameter> parameters) {
         super(NodePosition.ZERO);
         this.name = name;
         this.innerStmt = stmt;
         this.parameters = parameters == null ? ImmutableList.of() : parameters;
+        if (this.parameters != null) {
+            for (int i = 0; i < parameters.size(); i++) {
+                mysqlTypeCodes.add(0);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/PrepareStmt.java:43]: What's the meaning of '0' here?

Comment (liubotao) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/PrepareStmt.java:43]: only for placeholder initialization

**Comment by LiShuMing** [comment]
@mergify rebase

**Review by LiShuMing — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**